### DISCUSSION TITLE
Using multiple databases hustle

### DISCUSSION THREAD

**Discussion by demetris-manikas** [mikro-orm/nestjs#186] [General]
Hi!

I tried accessing two databases in my nestjs project using micro-orm (great by the way).

I spent a lot of time trying to figure out why using the documentation proposed way did not work for me.

When specifying a contextName for both database modules I was always getting weird errors about not having registered entities.

In the end after reading a post on how the same thing was implemented by nestjs/typeorm I tried to NOT set the contextName in one of the database modules and everything went smoothly after that.

If I am not missing something here I would suggest that this should be made clear in the docs that one of the database modules should be the default one (without a contextName defined) so that others are not affected as I was.

Thanks.

**Comment by B4nan** [comment]
It's hard to improve something when you don't say what is actually wrong. You are **not** supposed to use one instance without the context name in such setup. This sounds like you did not use the decorators property (since you need to put the context name in each ORM related decorator in that setup). 

[The docs in the readme here](https://github.com/mikro-orm/nestjs#multiple-database-connections) should be correct, if you think otherwise, please provide a failing repro.

**Reply by demetris-manikas** [reply]
I will if I get the time to.
Thanks for your prompt answer.

**Comment by demetris-manikas** [comment]
Hi. 
I created [here](https://github.com/demetris-manikas/micro-orm-test) a stripped down version of what I was trying to do.
After cloning run `docker compose up` and it will fail,
Then remove the 'postgres' contextName from the database module and from the relevant controller and module and it will rise.

Thanks for your time.

**Comment by demetris-manikas** [comment]
Hi. Did you get to look into this?
Should I refile an issue?

**Reply by B4nan** [reply]
Sorry, I was out for the weekend and was swamped on Friday and today with work stuff. Only checked the code briefly and didn't see anything wrong, let me give it a closer look.

**Reply by B4nan** [reply]
The problem here is the use of `useFactory` with a parameter, the ORM adapter does not know how to resolve it, see #184. 

Also you are missing `registerRequestContext: false` in the other `forRoot` call [here](https://github.com/demetris-manikas/micro-orm-test/blob/main/backend/src/database.module.ts#L34), as well as the `forMiddleware` call.

**Comment by demetris-manikas** [comment]
Thanks a lot for taking the time to look into this. 

Still my initial request holds. 
These things must be documented.

More people will eventually call useFactory() with a parameter. 

And reading this
```
You can define multiple database connections by registering multiple MikroOrmModule and setting their contextName. 

If you want to use middleware request context you must disable automatic middleware and register MikroOrmModule with forMiddleware() or use NestJS Injection Scope
```
I understood that setting registerRequestContext=false and forMiddleware should be used 

`if I want to use mikro-orm in middleware` . 

I will stick with only naming just one of the two since it is more convenient and it seems to work without the workarounds discused in #184 and without extra parameters definitions

```ts
MikroOrmModule.forRootAsync({
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: (configService: ConfigService) => {
        return {
          autoLoadEntities: true,
          driver: PostgreSqlDriver,
          host: configService.get('EMPOWER_POSTGRES_HOST'),
          port: +configService.get<number>('EMPOWER_POSTGRES_PORT', 5432),
          user: configService.get('EMPOWER_POSTGRES_USER'),
          password: configService.get('EMPOWER_POSTGRES_PASSWORD'),
          dbName: configService.get('EMPOWER_POSTGRES_DB'),
          extensions: [Migrator],
        };
      },
    }),

    MikroOrmModule.forRootAsync({
      imports: [ConfigModule],
      inject: [ConfigService],
      contextName: 'empower-eclass',
      useFactory: (configService: ConfigService) => {
        return {
          autoLoadEntities: true,
          driver: MariaDbDriver,
          host: configService.get('EMPOWER_MARIADB_HOST'),
          port: +configService.get<number>('EMPOWER_MARIADB_PORT', 5432),
          user: configService.get('EMPOWER_MARIADB_USER'),
          password: configService.get('EMPOWER_MARIADB_PASSWORD'),
          dbName: configService.get('EMPOWER_MARIADB_DB'),
        };
      },
    }),
  ],
```

**Reply by B4nan** [reply]
First of all you should understand this is a community-driven package, I did not introduce this feature myself, I haven't written the docs you don't like, nor have I started with this project. I am not even using nest myself, I just took over because the original maintainer went silent and I wanted to have something up to date. 

Second, your initial request is about multiple databases, while your actual problem has nothing to do with them.

So PR welcome, docs can be always better, and I am sure there are better and more "nest native" ways to deal with the things we see. Be the change.

**Reply by demetris-manikas** [reply]
I would love to be the change in many things, still my time is limited.
Anyway the whole situation is recorded here for future reference.
Thanks again for your time.

**Reply by B4nan** [reply]
I gave this another look, at this is not even about the `useFactory` problem I thought of. Your config service is provided correctly into the `useFactory`, but it's empty. You get a validation error that points to the actual problem, a missing db name, and that's because `configService.get('POSTGRES_DB')` returns `undefined`. I had to add a `.env` file and comment out [this line](https://github.com/demetris-manikas/micro-orm-test/blob/main/backend/src/app.module.ts#L10) to have it pass.

**Reply by demetris-manikas** [reply]
Consider this solved.

**Reply by B4nan** [reply]
The point is you are [ignoring it explicitly](https://github.com/demetris-manikas/micro-orm-test/blob/main/backend/src/app.module.ts#L10).

**Reply by demetris-manikas** [reply]
Let me check.

**Reply by demetris-manikas** [reply]
I double checked it. 
The problem lies on that I have to add the `registerRequestContext: false` property to both my setups. And I got confused by the docs because I understood that I have to set it only if I want to use the databases in middleware.
The ConfigService ignores the .env but the environment variables are set in the docker-compose file.

**Reply by demetris-manikas** [reply]
I updated the https://github.com/demetris-manikas/micro-orm-test repo.
Now it starts fine. If you remove the `registerRequestContext: false` property then it fails to start.
If you go on to remove the contextName from one of the two from the database module and the respective modules and services as well  then it again loads fine
